# Primeira Avaliação de Física Computacional I

* Aluno : Gabriel Wendell Celestino Rocha
* Matrícula :  20190112149 

* Disciplina : Física Computacional I
* Professor : Leonardo Machado
* Data de Entrega : 01/10/2020

In [2]:
import numpy as np

## Problema 5.

### Letra a)

Desejamos resolver a integral abaixo usando os métodos do trapézio, Simpson e Euler-Maclaurin, respectivamente. Para isso, vamos definir uma função `f(x)` que retornará a função a ser integrada.

$$\int_{0}^{2}(x^{5}-2x^{2}+1)dx$$

In [3]:
def f(x):
    return (x**5)-2*(x**2)+1

Vamos agora definir duas funções chamadas `trap(a,b,N)` e `simpson(a,b,N)` que irão resolver a nossa integral utilizando os métodos do trapézio e Simpson, respectivamente:

In [4]:
def trap(a,b,N):
    h = (b-a)/N

    integral = 0.5*(f(a)+f(b))

    for k in range(1,N):
        integral = integral + f(a+k*h)
    
    return h*integral

In [5]:
def simpson(a,b,N):
    h = (b-a)/N
    
    odd_sum = 0
    for k in range(1, N, 2):
        odd_sum += f(a+k*h)

    even_sum = 0
    for k in range(2, N, 2):
        even_sum += f(a+k*h)

    return (1/3)*h*(f(a)+f(b)+4*odd_sum+2*even_sum)

Agora vamos tentar implementar o método de Euler-Maclaurin para resolver a nossa integral. O método de EUler-Maclaurin nos diz que podemos resolver uma integral da seguinte forma:

$$\int_{a}^{b}f(x)dx=h\Bigg[\frac{1}{2}f(a)+\frac{1}{2}f(b)+\sum_{k=1}^{N-1}f(a+kh)\Bigg]+\frac{1}{12}h^{2}[f'(a)-f'(b)]+O(h^{4})$$

Note que teremos que trabalhar com as derivadas da função nos extremos do intervalo. Para isso, vamos agora criar mais duas funções, `deriv(x,h)` e `w(k,N)` que irão calcular as derivadas utilizando o método das diferenças centradas em $a$ e em $b$ e os pesos que usaremos no método de Euler-Maclaurin, respectivamente.

In [6]:
def w(k,N):
    if k==0 or k==N:
        return 0.5
    else:
        return 1

In [7]:
def deriv(x):
    h=10**-5
    return (f(x+(h/2)) - f(x-(h/2)))/h

Por fim, vamos criar uma última função chamada `euler_mac(a,b,N)` que irá implementar o método de Euler-Maclaurin

In [8]:
def euler_mac(a,b,N):
    h = (b-a)/N
    
    integral = 0
    for k in range(N+1):
        integral += w(k,N)*f(a+(k*h))

    integral = integral*h
    integral += ((h**2)/12)*(deriv(a)-deriv(b))
    
    return integral

Agora vamos chamar as nossas três funções e resolver a nossa integral por cada um dos métodos. Vamos usar um número $N=10$ de fatias

In [9]:
print('Método do trapézio: ', trap(0,2,10))
print('Método de Simpson: ', simpson(0,2,10))
print('Método de Euler-Maclaurin: ', euler_mac(0,2,10))

Método do trapézio:  7.572800000000001
Método de Simpson:  7.335466666666668
Método de Euler-Maclaurin:  7.332800000000797


Vamos resolver analiticamente a nossa integral para podermos comparar os resultados obtidos por cada método:

$$\int_{0}^{2}(x^{5}-2x^{2}+1)dx=\frac{x^{6}}{6}\Bigg|_{0}^{2}-\frac{2x^{3}}{3}\Bigg|_{0}^{2}+x|_{0}^{2}=\frac{2^{6}}{6}-\frac{2\cdot2^{3}}{3}+2=7.333\ldots=7.\overline{3}$$

$\square$

### Letra b)

Com base nos resultados expostos no item anterior, podemos notar que o método de Euler-Maclaurin e o método de Simpson nos fornecem resultados muito próximos um do outro, são faceis de se implementar e nos permitem calcular o erro em nossas integrais. 

Entretanto, esses resultados só podem ser obtidos caso tenhamos uma fórmula fechada para o nosso integrando, e na maioria das vezes isso não acontece. Um exemplo nítido seria um onde o integrando não é uma função conhecida mas sim um conjunto de dados coletados em laboratória através de uma série de experimentos. 

Como na maioria das vezes tanto na Física quanto na Matemática não se conhece a função a ser integrada, isso justificaria o pouco uso tanto da regra de Simpson quanto da regra de Euler-Maclaurin.

$\square$